Refactor and enhance `TaskPluginTrait`

ditsuke opened the pull request:
### Summary of Changes
- Adds a standard routine handler.
- Upgrades `enhanceTaskItemForm()` to a complete event handler.
- Renames some methods.
- Improves documentation.
- A new status code.
- Code style improvements.

#### TODO
- [ ] Update task plugins. At the moment they're not compatible with these changes.

### Testing Instructions
--


### Actual result BEFORE applying this Pull Request
--


### Expected result AFTER applying this Pull Request
--


### Documentation Changes Required
--

ditsuke committed:
Add standard routine handler

`standardRoutineHandler()` can take care of initialising the routine,
calling the associated callable and ending the routine all without
any logic in the plugin class if each routine has a corresponding
callable.

ditsuke committed:
Fix copyright formatting

ditsuke committed:
Update Task\Status

Adds status code for invalid return and improves member doc blocks.

ditsuke committed:
Refactor TaskPluginTrait

- Renames methods:
 - `taskLog()` => `logTask()`
 - `taskEnd()` => `endRoutine()`
 - `taskStart()` => `initRoutine()`
- Upgrade `enhanceTaskItemForm()` to a complete event handler. For
  plugins targeting only the task form, this method can now be mapped
  straight to the `onContentPrepareForm` event.
- Code style improvements and fixes for global namespace qualification
  on some `\Exception` references.
- More explicit checks on variables.
- `Event` type arguments are now `EventInterface` so other
  implementations stay compatible.
- Removes deprecated logging nd `$log` param from `endRoutine()`.

ditsuke committed:
Improve TaskPluginTrait documentation

Adds and improves on the information in the doc blocks with context,
usage hints and related information. Also updates some parameter
information and fixes the tag order (`@since` <=> `@throws`).

bembelimen commented on administrator/components/com_scheduler/src/Traits/TaskPluginTrait.php:
```diff
@@ -20,128 +18,156 @@
+
+		if ($form->getName() !== 'com_scheduler.task')
+		{
+			return false;
```
I guess, nothing wrong with not beeing a ```com_scheduler.task```, so it shouldn't be ```false``` but ```true``` (otherwise something above could stop/break?)

bembelimen commented on administrator/components/com_scheduler/src/Traits/TaskPluginTrait.php:
```diff
@@ -20,128 +18,156 @@
+		// Return if routine is not supported by the plugin or the routine does not have a form linked in TASKS_MAP.
+		if (!$isSupported || empty($enhancementFormName))
 		{
 			return false;
```
Here again I would return ```true```

bembelimen commented on administrator/components/com_scheduler/src/Traits/TaskPluginTrait.php:
```diff
@@ -221,4 +248,59 @@ protected function addTaskLog(string $message, string $priority = 'info'): void
+		{
+			$exitCode = \call_user_func($callable);
+		}
+		elseif (!empty($callable) && method_exists($this, $callable))
```
Please check, if e.g. private methods gives ```true``` in this check, but fail when called two lines below and if ```is_callable``` fix this then?

ditsuke replied:
Using now reflection to get around visibility and enforce signatures.

bembelimen commented on administrator/components/com_scheduler/src/Traits/TaskPluginTrait.php:
```diff
@@ -20,128 +18,156 @@
 
-		if (!$isSupported || !$enhancementForm = self::TASKS_MAP[$routineId]['form'] ?? '')
+		// Return if routine is not supported by the plugin or the routine does not have a form linked in TASKS_MAP.
+		if (!$isSupported || empty($enhancementFormName))
```
I know, it's a stupid use case, but if the file is called ```0.xml``` this check would do a false negative. Better use ```strlen```?

ditsuke replied:
Good idea

bembelimen commented on administrator/components/com_scheduler/src/Traits/TaskPluginTrait.php:
```diff
@@ -20,128 +18,156 @@
-		$path = dirname((new \ReflectionClass(static::class))->getFileName());
+		// We expect the form XML in "{PLUGIN_PATH}/forms/{FORM_NAME}.xml"
+		$path                = dirname((new \ReflectionClass(static::class))->getFileName());
+		$enhancementFormFile = $path . '/forms/' . $enhancementFormName . '.xml';
```
You can secure "user input paths" via ```Path::check```, to prevent ```../../```

ditsuke replied:
done

bembelimen commented on administrator/components/com_scheduler/src/Traits/TaskPluginTrait.php:
```diff
@@ -221,4 +248,59 @@ protected function addTaskLog(string $message, string $priority = 'info'): void
+		$callable  = self::TASKS_MAP[$routineId]['call'] ?? '';
+		$exitCode  = Status::NO_EXIT;
+
+		if (!empty($callable) && is_callable($callable))
```
I'm not sure, if we should allow the developer to call any function existing in the environment. I think we should only allow calling things within the plugin (like you do below and if they want to call functions, they can do it there)

ditsuke replied:
Thanks. Done with [`c521a5b` (#29)](https://github.com/joomla-projects/soc21_website-cronjob/pull/29/commits/c521a5ba4551c6b67d1b9be2e03c1ae22e261ff5)

bembelimen commented on administrator/components/com_scheduler/src/Traits/TaskPluginTrait.php:
```diff
@@ -221,4 +248,59 @@ protected function addTaskLog(string $message, string $priority = 'info'): void
+		elseif (!is_integer($exitCode))
+		{
+			$exitCode = Status::INVALID_EXIT;
+		}
```
I think, this whole check could be a bit improved, there could be exitCodes which are not supported at all.

ditsuke replied:
We check against `Task\Status` now with [`c521a5b` (#29)](https://github.com/joomla-projects/soc21_website-cronjob/pull/29/commits/c521a5ba4551c6b67d1b9be2e03c1ae22e261ff5)

bembelimen commented on administrator/components/com_scheduler/src/Traits/TaskPluginTrait.php:
```diff
@@ -221,4 +248,59 @@ protected function addTaskLog(string $message, string $priority = 'info'): void
+		}
+		elseif (!empty($callable) && method_exists($this, $callable))
+		{
+			$exitCode = \call_user_func([$this, $callable]);
```
Thought about a try/catch if really something goes wrong?

ditsuke replied:
done!

bembelimen commented on administrator/components/com_scheduler/src/Traits/TaskPluginTrait.php:
```diff
@@ -20,128 +18,156 @@
 	 * @since  __DEPLOY_VERSION__
 	 */
-	private function taskStart(ExecuteTaskEvent $event): void
+	protected function initRoutine(ExecuteTaskEvent $event): void
```
if you have ```endRoutine``` it should be probably ```startRoutine```?

ditsuke replied:
done

bembelimen commented on administrator/components/com_scheduler/src/Traits/TaskPluginTrait.php:
```diff
@@ -20,128 +18,156 @@
 	 * @var array
 	 * @since  __DEPLOY_VERSION__
 	 */
 	protected $snapshot = [];
```
TBH, that's just me, but I'm not 100% happy with the naming of this variable and that it is an array (snapshots are for me something I could recover to (e.g. backup) and the array is very fragile)

ditsuke replied:
Good point, I've been thinking for a while to turn it into a `Task\RoutineSnapshot` or similar, or perhaps inject the snapshot from the Task into the routine.

bembelimen commented on administrator/components/com_scheduler/src/Traits/TaskPluginTrait.php:
```diff
@@ -1,14 +1,12 @@
 <?php
```
Your trait needs an interface

bembelimen commented on administrator/components/com_scheduler/src/Traits/TaskPluginTrait.php:
```diff
@@ -159,47 +185,48 @@ public function advertiseRoutines(Event $event): void
+		 * Depending on when the form is loaded, the ID may either be in $data or the data already bound to the form.
+		 * $data can also either be an object or an array.
 		 */
 		$routineId = $data->taskOption->type ?? $data->type ?? $data['type'] ?? $form->getValue('type') ?? $data['taskOption']->type;
```
```$data['taskOption']->type``` is always set?

ditsuke committed:
Implement suggestions on TaskPluginTrait

- Refactors `initRoutine()` => `startRoutine()`
- Return true for irrelevant enhanceTaskItemForm() contexts.
- Path checks for routine forms.
- Allow only class methods with `standardRoutineHandler()`
- Adds signature checks on standard routine methods.
- Adds validation on routine return codes against `Task\Status`.

ditsuke committed:
Update TaskPluginTrait::standardRoutineHandler()

Now enforces a single required parameter of type `ExecuteTaskEvent`
(from `EventInterface`).

ditsuke committed:
Update task plugins

- Remove methods now offered by TaskPluginTrait.
- The 'call' TASKS_MAP param is now 'method'.
- Update event subscriptions.
- Update docs.
- Improve code-style and copyright blocks.

ditsuke committed:
Update TaskPluginTrait doc

Updates signature reference for standard routine methods.
